### PR TITLE
Invert foreground color when background is light

### PR DESCRIPTION
**Pull request by sangwoo108**
Resolves https://github.com/brave/brave-browser/issues/25392
Resolves https://github.com/brave/brave-browser/issues/25393

For readability, the foreground color is inverted when the background is light. This is based on WCGA2.0 spec.

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### Automated
test-unit ColorUtil.test.ts

### Manual
Even when using light solid color background, we should be read stats UI easily.

### Storybook
* Go to "regular"
* opt out "sponsored image" on right side control panel.
* try changing background from "defaultImage" on the right side control panel.

### CODE REVIEW

File: components/brave_new_tab_ui/components/default/stats/style.ts [hunk 1]
```diff
@@ -56,10 +52,10 @@ export const StyledStatsItemText = styled('span')<{}>`
   letter-spacing: 0;
 `
 
-export const StyledStatsItemDescription = styled('div')<{}>`
+export const StyledStatsItemDescription = styled('div')<{overriddenTextColor?: string}>`
   font-size: 16px;
   font-weight: 500;
-  color: #FFFFFF;
+  ${p => !p.overriddenTextColor && css`color: #FFFFFF;`}
```
Comment (fallaciousreasoning) [inline components/brave_new_tab_ui/components/default/stats/style.ts:58]: Optional: I think falsy values are removed by styled-components, so you could also write this as

```suggestion
  color: ${p => !p.overriddenTextColor && '#FFFFFF;'}
```

File: components/brave_new_tab_ui/components/default/widget/assets/ellipsis.ts [hunk 2]
```diff
@@ -3,13 +3,6 @@
...
 import EllipsisIcon from '../../../popupMenu/ellipsisIcon'
 
-interface Props {
-  lightWidget?: boolean
-}
-
-export default styled(EllipsisIcon)<Props>`
-  color: ${p => p.lightWidget ? '#495057' : '#ffffff'};
-`
+export default EllipsisIcon
```
Comment (fallaciousreasoning) [inline components/brave_new_tab_ui/components/default/widget/assets/ellipsis.ts:8]: Do we even need this file? It seems kinda pointless :laughing:

File: components/brave_new_tab_ui/containers/newTab/index.tsx [hunk 3]
```diff
@@ -1158,6 +1158,11 @@ class NewTabPage extends React.Component<Props, State> {
       cryptoContent = null
     }
 
+    let overriddenForegroundColor: string | undefined
```
Comment (fallaciousreasoning) [inline components/brave_new_tab_ui/containers/newTab/index.tsx:1161]: Optional:

```suggestion
    const overriddenForegroundColor = !newTabData.brandedWallpaper && newTabData.backgroundWallpaper?.type === 'color' ? newTabData.backgroundWallpaper.overriddenForegroundColor : undefined
```

If we won't have a `overriddenForegroundColor` in the case we don't need it we could write 
```suggestion
// string | undefined
const overriddenForegroundColor = newTabData.backgroundWallpaper?.overridenForegroundColor
```

File: components/brave_new_tab_ui/containers/newTab/index.tsx [hunk 4]
```diff
@@ -1247,12 +1253,13 @@ class NewTabPage extends React.Component<Props, State> {
...
               onClickSettings={this.openSettings}
               onDismissBraveTalkPrompt={this.props.actions.dismissBraveTalkPrompt}
+              color={overriddenForegroundColor ?? '#ffffff'}
             />
             </Page.FooterContent>
           </Page.Footer>
           {newTabData.showToday &&
           <Page.GridItemNavigationBraveToday>
-            <BraveTodayHint />
+            <BraveTodayHint color={overriddenForegroundColor ?? 'white'}/>
```
Comment (fallaciousreasoning) [inline components/brave_new_tab_ui/containers/newTab/index.tsx:1262]: Nit: Make this use a hex color, like the other ones
```suggestion
            <BraveTodayHint color={overriddenForegroundColor ?? '#ffffff'}/>
```

File: components/brave_new_tab_ui/components/default/page/index.tsx [hunk 5]
```diff
@@ -461,19 +465,20 @@ export const IconButtonSideText = styled('label')<IconButtonSideTextProps>`
...
   textDirection: string
+  color: string
 }
 
 export const IconButtonContainer = styled('div')<IconButtonContainerProps>`
   font-family: ${p => p.theme.fontFamily.heading};
   font-size: 13px;
   font-weight: 600;
-  color: rgba(255,255,255,0.8);
+  color: ${p => p.color + 'CC' /* == 0.8 */};
```
[the remark this answers is not in the document]
Comment (fallaciousreasoning) [reply components/brave_new_tab_ui/components/default/page/index.tsx:475]: Could we potentially use the `ColorUtils` to parse the color as an RGB value and then set the alpha channel?

File: components/brave_new_tab_ui/helpers/colorUtil.ts [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+        return 0.0
+    }
+
+    const cachedLuminance = this.cachedLuminance
+    if (cachedLuminance) {
+      return cachedLuminance
+    }
+
+    // Step 1. Lerp each channel to a value between 0 and 1.
+    [r, g, b] = [r, g, b].map(v => v / 255);
```
Comment (fallaciousreasoning) [inline components/brave_new_tab_ui/helpers/colorUtil.ts:98]: This is beautiful :smile:

File: components/brave_new_tab_ui/helpers/colorUtil.ts [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+
+/**
+ * Convenient function to get readability for our background type
+ * @param {NewTab.BackgroundWallpaper | undefined} background -
+ * NewTab.BackgroundWallpaper | undefined
+ * @returns {boolean} - true if background is readable
+ */
+export default function isReadableOnBackground (background: NewTab.BackgroundWallpaper | undefined) {
+  if (!background || background.type !== 'color') {
+    // Consider other cases readable. We can't determine the color from images.
```
Comment (fallaciousreasoning) [inline components/brave_new_tab_ui/helpers/colorUtil.ts:160]: It'd be cool if we could do something like get the average darkness or something. Probably way outside the scope of this PR though :laughing:

**Review by fallaciousreasoning — APPROVED** [review]
Okay, so this looks pretty good to me but I had an idea as I was reading the last bit of the PR: What if instead of adding all of these JS props we optionally added a 

```
--override-readability-color-rgb: 0, 0, 0;
--override-readability-color: rgb(var(--override-readability-color-rgb));
```

css variable. Then each component that might be overridden could do this:

```
styled.whatever`
    // Fall back to white
    color: var(--override-readability-color, #ffffff);
`

styled.other`
    // Fall back to --interactive9
   color: var(--override-readability-color, var(--interactive9));
`

styled.withAlpha`
    // Fallback to white, either with 0.8 opacity
    color: rgba(var(--override-readability-color-rgb, 255, 255, 255), 0.8);
`
```

Then we wouldn't have to pass these JS variables around all over the place!

You could set the variables in `components/brave_new_tab_ui/containers/newTab/index.tsx`

```tsx

useEffect(() => {
	if (currentColorsAreReadable) return;

    const style = document.documentElement.style;
	style.setProperty('--override-readability-color-rgb', '0, 0, 0');
	style.setProperty('--override-readability-color', 'rgb(var(--override-readability-color))');
    return () => {
		style.removeProperty('--override-readability-color-rgb')
		style.removeProperty('--override-readability-color')
    }
}, [background, otherThingsThatChangeTheBackground]);

```

File: components/brave_new_tab_ui/components/default/page/index.tsx [hunk 8]
```diff
@@ -461,19 +465,20 @@ export const IconButtonSideText = styled('label')<IconButtonSideTextProps>`
...
   margin-right: ${p => p.textDirection === 'ltr' && '8px'};
   margin-left: ${p => p.textDirection === 'rtl' && '8px'};
-  border-right: ${p => p.textDirection === 'ltr' && '1px solid rgba(255, 255, 255, 0.6)'};
-  border-left: ${p => p.textDirection === 'rtl' && '1px solid rgba(255, 255, 255, 0.6)'};
+  border-right: ${p => p.textDirection === 'ltr' && `1px solid ${p.color + '99' /* == 0.6 */}`};
```
[the remark this answers is not in the document]
Comment (sangwoo108) [reply components/brave_new_tab_ui/components/default/page/index.tsx]: Oh, yeah. It looks awesome. I guess it's still in WD status, so not started implementing it yet?

File: components/brave_new_tab_ui/components/default/page/index.tsx [hunk 9]
```diff
@@ -467,13 +467,26 @@ export const IconButtonContainer = styled('div')<IconButtonContainerProps>`
...
+
+export const OverrideReadabilityColor = createGlobalStyle<{override: boolean}>`
+  body {
+    ${p => p.override && css`
+      --override-readability-color-rgb: 0, 0, 0;
+      --override-readability-color: rgb(0, 0, 0);
+      
+      // override color property in resets.css. Not sure why this happens but,
+      // the value in the stylesheet wins over variables above.
+      color: inherit
```
Comment (sangwoo108) [inline components/brave_new_tab_ui/components/default/page/index.tsx:489]: I couldn't figure out why this happens. This happens when we declare variables with setProperty(). I tried setting property for `body`, but it didn't work. As a workaround, I override `color` here.

The previous trial can be found at https://github.com/brave/brave-core/pull/15230/commits/27a0d34e36d85962ef0b7d71d98762f57421151a

Screen Shot 2022-09-28 at 4 59 33 PM

**Review by sangwoo108 — COMMENTED** [review]
Thanks @fallaciousreasoning ! I love the "variable idea" so much. But there was a weird thing I couldn't understand. I made an alternative way to stick to the idea. Could you please take a look again?

File: components/brave_new_tab_ui/components/default/page/index.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (sangwoo108) [reply components/brave_new_tab_ui/components/default/page/index.tsx]: > So this won't work if we pass in an RGB, HSL, CSS color, or CSS variable right?

Exactly. It's super fragile.

File: components/brave_new_tab_ui/helpers/colorUtil.ts (hunk 6 above)
Comment (sangwoo108) [reply components/brave_new_tab_ui/helpers/colorUtil.ts:98]: Thanks! I've always been jealous of python and javascript for this style.

File: components/brave_new_tab_ui/components/default/page/index.tsx (hunk 9 above)
Comment (fallaciousreasoning) [reply components/brave_new_tab_ui/components/default/page/index.tsx:489]: I think I like yours better anyway :smile: It's weird we need to set the `color: inherit` though - I just tested removing it and it seems to work without

```ts
css`
      --override-readability-color-rgb: 0, 0, 0;
      --override-readability-color: rgb(0, 0, 0);
`
```

File: components/brave_new_tab_ui/components/default/page/index.tsx [hunk 10]
```diff
@@ -467,13 +467,26 @@ export const IconButtonContainer = styled('div')<IconButtonContainerProps>`
...
+  border-left: ${p => p.textDirection === 'rtl' && 'rgba(var(--override-readability-color-rgb, 255, 255, 255), 0.6)'};
 
   &:hover {
-    color: #ffffff;
+    color: ${p => p.color};
+  }
+`
+
+export const OverrideReadabilityColor = createGlobalStyle<{override: boolean}>`
+  body {
```
Comment (fallaciousreasoning) [inline components/brave_new_tab_ui/components/default/page/index.tsx:482]: Optional nit: 
```suggestion
  :root {
```

I think we set most other CSS variables on the root, so we can do that with yours too

**Review by fallaciousreasoning — APPROVED** [review]
Still LGTM @sangwoo108 - I like your solution with `createGlobalStyle` more than my `useEffect` anyway - it's much more elegant.

File: components/brave_new_tab_ui/components/default/page/index.tsx [hunk 11]
```diff
@@ -467,13 +467,26 @@ export const IconButtonContainer = styled('div')<IconButtonContainerProps>`
...
   font-size: 13px;
   font-weight: 600;
-  color: rgba(255,255,255,0.8);
+  color: rgba(var(--override-readability-color-rgb, 255, 255, 255), 0.8);
   margin-right: ${p => p.textDirection === 'ltr' && '8px'};
   margin-left: ${p => p.textDirection === 'rtl' && '8px'};
-  border-right: ${p => p.textDirection === 'ltr' && '1px solid rgba(255, 255, 255, 0.6)'};
-  border-left: ${p => p.textDirection === 'rtl' && '1px solid rgba(255, 255, 255, 0.6)'};
+  border-right: ${p => p.textDirection === 'ltr' && 'rgba(var(--override-readability-color-rgb, 255, 255, 255), 0.6)'};
+  border-left: ${p => p.textDirection === 'rtl' && 'rgba(var(--override-readability-color-rgb, 255, 255, 255), 0.6)'};
```
[the remark this answers is not in the document]
Comment (sangwoo108) [reply components/brave_new_tab_ui/components/default/page/index.tsx:474]: Ooops, thanks!

File: components/brave_new_tab_ui/components/default/widget/widgetMenu.tsx [hunk 12]
```diff
@@ -32,8 +32,8 @@ interface Props {
   onAddSite?: () => void
   customLinksEnabled?: boolean
   onToggleCustomLinksEnabled?: () => void
-  lightWidget?: boolean
   paddingType: 'none' | 'right' | 'default'
+  color?: string
```
[the remark this answers is not in the document]
Comment (sangwoo108) [reply components/brave_new_tab_ui/components/default/widget/widgetMenu.tsx:36]: oops, removed.

File: components/brave_new_tab_ui/components/default/page/index.tsx (hunk 10 above)
Comment (sangwoo108) [reply components/brave_new_tab_ui/components/default/page/index.tsx:482]: Yep that's better! thanks.

File: components/brave_new_tab_ui/components/default/page/index.tsx (hunk 9 above)
Comment (sangwoo108) [reply components/brave_new_tab_ui/components/default/page/index.tsx:489]: Ah yeah, I found the culprit. I used wrong variable for some UI.

```diff
-  color: var(--override-readability-color-rgb, #FFFFFF);  <--! oops, rgb is 0, 0, 0 -->
+  color: var(--override-readability-color, #FFFFFF);
```

**Comment by sangwoo108** [comment]
Thank you so much @fallaciousreasoning ! I learned a lot from you as always!

**Pull request merged**